File: src/core/GameDirector.java
package core;

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import org.newdawn.slick.AppGameContainer;
import org.newdawn.slick.Color;
import org.newdawn.slick.SlickException;
import org.newdawn.slick.state.transition.FadeInTransition;
import org.newdawn.slick.state.transition.FadeOutTransition;
import org.newdawn.slick.state.transition.RotateTransition;
import org.newdawn.slick.state.transition.Transition;

import scene.*;
import scene.encounter.*;
import scene.test.*;

import model.*;

/**
 * Directs the logical functionality of the game. Sets everything in motion.
 */

public class GameDirector implements SceneListener {
	private static GameDirector sharedDirector;
	
	private FontStore fontManager;
	
	private SceneDirector sceneDirector;
	private AppGameContainer container;
	
	private Game game;
	private WorldMap worldMap;
	/**
	 * Constructs a game director object.
	 */
	private GameDirector() {
		 sharedDirector = this;
		 
		 //fontManager = new FontManager();

		 sceneDirector = new SceneDirector("Oregon Trail");
		 worldMap = new WorldMap();
		 game = new Game(worldMap);
	}
	
	/**
	 * Singleton Design pattern. Returns the single instance of game director object that drives the game, for the scenes to interact with.
	 * @return handle of game director for scene to interact with
	 */
	public static SceneListener sharedSceneListener() {
		return sharedDirector;
	}

	/**
	 * Launches window that contains game.
	 */
	public void start() {
		try {
			container = new AppGameContainer(sceneDirector);
			container.setDisplayMode(1024, 576, false);
			container.setTargetFrameRate(60);
			container.setAlwaysRender(true);
			container.setShowFPS(false);
			container.start();
		} catch (SlickException e) {
			e.printStackTrace();
		}
	}
	
	/**
	 * Returns the game object.
	 * @return the game
	 */
	public Game getGame() {
		return game;
	}
	
	/**
	 * Returns the container that holds the gui element of the game.
	 * @return the game container
	 */
	public AppGameContainer getContainer() {
		return container;
	}
	
	/**
	 * Determines the appropriate game scene to return based on a passed id.
	 * @param id the {@code SceneID} desired
	 * @return the handle to the newly created scene
	 */
	private Scene sceneForSceneID(SceneID id) {
		switch (id) {
		case MAINMENU:
			return new MainMenuScene();
		case PARTYCREATION:
			return new PartyCreationScene(game.getPlayer());
		case TOWN:
			game.resetStoreInventory(worldMap.getCurrLocationNode());
			return new TownScene(game.getPlayer().getParty(), worldMap.getCurrLocationNode());
		case STORE:
			return new StoreScene(game.getPlayer().getParty(), game.getStoreInventory(), (1 + worldMap.getCurrLocationNode().getRank()/10));
		case PARTYINVENTORY:
			return new PartyInventoryScene(game.getPlayer().getParty());
		case SCENESELECTOR:
			SoundStore.get().stop();
			return new SceneSelectorScene(game);
		case HUNT:
			return new HuntScene(game.getPlayer().getParty());
		case TRAIL:
			return new TrailScene(game.getPlayer().getParty(), new RandomEncounterTable(getEncounterList()));
		case GAMEOVER:
			return new GameOverScene();
		case VICTORY:
			return new VictoryScene();
		case COMPONENTTEST:
			return new ComponentTestScene();
		case TRAILTEST:
			return new TrailTestScene(game.getPlayer().getParty());
		case MAP:
			return new MapScene(game.getWorldMap());
		case RIVER:
			return new RiverScene(game.getPlayer().getParty());
		}
		
		return null;
	}

	/*----------------------
	  SceneDelegate
	  ----------------------*/
	@Override
	public void requestScene(SceneID id, Scene lastScene, boolean popLastScene) {
		Transition outTransition = null;
		Transition inTransition = null;
		
		if (game.getPlayer().getParty() != null && game.getPlayer().getParty().getLocation() != null) {
			worldMap.setCurrLocationNode(game.getPlayer().getParty().getLocation());
			worldMap.setCurrTrail(game.getPlayer().getParty().getTrail());
		}
		
		Scene newScene = id != null ? sceneForSceneID(id): null;
		
		if (worldMap.getCurrLocationNode().getRank() == worldMap.getMaxRank()) {
			newScene = new VictoryScene();
		}
		
		if (newScene instanceof VictoryScene || newScene instanceof GameOverScene) {
			inTransition = new RotateTransition(Color.black);
		} else if (newScene instanceof PartyInventoryScene && lastScene instanceof StoreScene) {
			// Requested Party Inventory Scene
			newScene = new PartyInventoryScene(game.getPlayer().getParty(), ((StoreScene) lastScene).getInventory());
		} else if (lastScene instanceof PartyCreationScene) {
			// Last scene was Party Creation Scene
			game.getPlayer().getParty().setLocation(game.getWorldMap().getMapHead());
		}
		
		if (newScene != null) {
			lastScene.disable();
			
			if (outTransition == null) {
				outTransition = new FadeOutTransition(Color.black);
			}
			if (inTransition == null) {
				inTransition = new FadeInTransition(Color.black);
			}
			
			sceneDirector.pushScene(newScene, popLastScene, true, outTransition, inTransition);
		}
	}
	
	@Override
	public void sceneDidEnd(Scene scene) {
		scene.disable();
		sceneDirector.popScene(true);
	}
	
	@Override
	public FontStore getFontManager() {
		return fontManager;
	}
	
	@Override
	public void showSceneSelector() {
		sceneDirector.replaceStackWithScene(sceneForSceneID(SceneID.SCENESELECTOR));
	}
	
	/**
	 * Gets the encounter list to populate the trail scenes random encounter table
	 * @return The mapping of scene types to probabilities
	 */
	private List<Encounter> getEncounterList() {
		Random rand = new Random();
		List<Encounter> encounterList = new ArrayList<Encounter>();
		
		for ( EncounterID encounter : EncounterID.values() ) {
			encounterList.add(EncounterID.getEncounter(game.getPlayer().getParty(), encounter, 5 * (rand.nextInt(10)+1)));
		}
		return encounterList;
	}
	
	public void serialize() {
		try {
			FileOutputStream fileOut = new FileOutputStream("resources/serialized/game.ser");
			ObjectOutputStream out = new ObjectOutputStream(fileOut);
			out.writeObject(game);
			out.close();
			fileOut.close();
		} catch (IOException i) {
			i.printStackTrace();
		}
	}
	
	@Override
	public Game deserialize() {
		try {
			FileInputStream fileIn = new FileInputStream("resources/serialized/game.ser");
			ObjectInputStream in = new ObjectInputStream(fileIn);
			game = (Game) in.readObject();
			in.close();
			fileIn.close();
		} catch (IOException i) {
			i.printStackTrace();
			return null;
		} catch (ClassNotFoundException e) {
			Logger.log("Game class not found", Logger.Level.INFO);
			e.printStackTrace();
			return null;
		}
		return game;
	}

	/*----------------------
	  Main
	  ----------------------*/
	/**
	 * The game main method
	 * @param args
	 */
	public static void main(String[] args) {
		Logger.log("-----------------NEW GAME STARTED-----------------", Logger.Level.INFO);
		GameDirector gameDirector = new GameDirector();
		gameDirector.start();
	}
}

File: web/src/core/GameDirector.ts
import { SceneDirector } from './SceneDirector';
import { InputManager } from './InputManager';
import { SplashScene } from '../scene/SplashScene';
import { LoadingScene } from '../scene/LoadingScene';
import { MainMenuScene } from '../scene/MainMenuScene';
import { PartyCreationScene } from '../scene/PartyCreationScene';
import { MapScene } from '../scene/MapScene';
import { GameOverScene } from '../scene/GameOverScene';
import { VictoryScene } from '../scene/VictoryScene';
import { OptionsScene } from '../scene/OptionsScene';
import { Scene, SceneDelegate } from '../scene/Scene';
import { SceneID } from '../scene/SceneID';
import { Game } from '../model/Game';
import { WorldMap } from '../model/WorldMap';
import { FadeOutTransition, FadeInTransition, RotateTransition } from './Transition';
import { Color } from './Color';

export class GameDirector implements SceneDelegate {
  private canvas: HTMLCanvasElement;
  private ctx: CanvasRenderingContext2D;
  private sceneDirector: SceneDirector;
  private inputManager: InputManager;
  private lastTimestamp: number = 0;
  private running: boolean = false;
  private game: Game;
  private worldMap: WorldMap;

  constructor(canvas: HTMLCanvasElement) {
    this.canvas = canvas;
    const ctx = canvas.getContext('2d');
    if (!ctx) {
      throw new Error('Failed to get 2D context from canvas');
    }
    this.ctx = ctx;
    this.ctx.imageSmoothingEnabled = false;

    this.sceneDirector = new SceneDirector();
    this.inputManager = new InputManager(canvas);
    this.inputManager.setDelegate(this.sceneDirector);

    this.worldMap = new WorldMap(120);
    this.game = new Game(this.worldMap);

    // Expose requestScene to scenes via window
    (window as any).__requestScene = (id: SceneID, lastScene: Scene | null, replace: boolean) => {
      this.requestScene(id, lastScene, replace);
    };
  }

  start(): void {
    if (this.running) return;
    this.running = true;

    // Skip splash, go straight to loading
    const loadingScene = new LoadingScene(this.canvas.width, this.canvas.height, () => {
      const mainMenuScene = this.sceneForSceneID(SceneID.MAINMENU);
      if (mainMenuScene) {
        this.sceneDirector.pushScene(mainMenuScene, true);
      }
    });

    this.sceneDirector.pushScene(loadingScene, false);
    this.lastTimestamp = performance.now();
    this.gameLoop(this.lastTimestamp);
  }

  private gameLoop(timestamp: number): void {
    if (!this.running) return;

    const delta = timestamp - this.lastTimestamp;
    this.lastTimestamp = timestamp;

    this.sceneDirector.update(delta);

    this.ctx.fillStyle = '#000000';
    this.ctx.fillRect(0, 0, this.canvas.width, this.canvas.height);

    this.sceneDirector.render(this.ctx);

    requestAnimationFrame((ts) => this.gameLoop(ts));
  }

  stop(): void {
    this.running = false;
  }

  private sceneForSceneID(id: SceneID, _lastScene?: Scene): Scene | null {
    let scene: Scene | null = null;

    switch (id) {
      case SceneID.SPLASH:
        scene = new SplashScene(this.canvas.width, this.canvas.height, () => {
          this.requestScene(SceneID.LOADING, null, true);
        });
        break;
      case SceneID.LOADING:
        scene = new LoadingScene(this.canvas.width, this.canvas.height, () => {
          this.requestScene(SceneID.MAINMENU, null, true);
        });
        break;
      case SceneID.MAINMENU:
        scene = new MainMenuScene(this.canvas.width, this.canvas.height);
        break;
      case SceneID.PARTYCREATION:
        scene = new PartyCreationScene(this.canvas.width, this.canvas.height);
        break;
      case SceneID.TOWN:
        this.game.resetStoreInventory(this.worldMap.getCurrLocationNode());
        // TownScene not yet implemented - placeholder
        console.log('TownScene requested but not yet implemented');
        break;
      case SceneID.STORE:
        // StoreScene not yet implemented - placeholder
        console.log('StoreScene requested but not yet implemented');
        break;
      case SceneID.PARTYINVENTORY:
        // PartyInventoryScene not yet implemented - placeholder
        console.log('PartyInventoryScene requested but not yet implemented');
        break;
      case SceneID.HUNT:
        // HuntScene not yet implemented - placeholder
        console.log('HuntScene requested but not yet implemented');
        break;
      case SceneID.TRAIL:
        // TrailScene in progress - placeholder
        console.log('TrailScene requested but not yet implemented');
        break;
      case SceneID.GAMEOVER:
        scene = new GameOverScene(this.canvas.width, this.canvas.height);
        break;
      case SceneID.VICTORY:
        scene = new VictoryScene(this.canvas.width, this.canvas.height);
        break;
      case SceneID.MAP:
        scene = new MapScene(this.game.getWorldMap());
        break;
      case SceneID.RIVER:
        // RiverScene not yet implemented - placeholder
        console.log('RiverScene requested but not yet implemented');
        break;
      case SceneID.OPTIONS:
        scene = new OptionsScene(this.canvas.width, this.canvas.height);
        (scene as OptionsScene).setOnMainMenu(() => this.resetToMainMenu());
        (scene as OptionsScene).setOnBack(() => {
          const currentScene = this.sceneDirector.currentScene();
          if (currentScene) {
            currentScene.leave();
          }
        });
        break;
      case SceneID.TAVERN:
        // TavernScene not yet implemented - placeholder
        console.log('TavernScene requested but not yet implemented');
        break;
      case SceneID.SCENESELECTOR:
        // SceneSelectorScene not yet implemented - placeholder
        console.log('SceneSelectorScene requested but not yet implemented');
        break;
    }

    if (scene) {
      scene.setSceneDelegate(this);
    }

    return scene;
  }

  requestScene(id: SceneID | null, lastScene: Scene | null, popLastScene: boolean): void {
    let outTransition: FadeOutTransition | RotateTransition | null = null;
    let inTransition: FadeInTransition | RotateTransition | null = null;

    if (
      this.game.getPlayer().getParty() &&
      this.game.getPlayer().getParty()!.getLocation()
    ) {
      this.worldMap.setCurrLocationNode(this.game.getPlayer().getParty()!.getLocation()!);
      const trail = this.game.getPlayer().getParty()!.getTrail();
      if (trail) {
        this.worldMap.setCurrTrail(trail);
      }
    }

    let newScene: Scene | null = id !== null ? this.sceneForSceneID(id, lastScene || undefined) : null;

    if (this.worldMap.getCurrLocationNode().getRank() === this.worldMap.getMaxRank()) {
      newScene = new VictoryScene(this.canvas.width, this.canvas.height);
    }

    if (newScene instanceof VictoryScene || newScene instanceof GameOverScene) {
      inTransition = new RotateTransition(Color.black);
    } else if (newScene && lastScene) {
      // Special case: PartyInventoryScene from StoreScene
      // Note: Requires StoreScene implementation

      // Special case: After PartyCreationScene
      const lastSceneID = lastScene.getID();
      if (lastSceneID === SceneID.PARTYCREATION) {
        this.game.getPlayer().getParty()!.setLocation(this.game.getWorldMap().getMapHead());
      }
    }

    if (newScene) {
      if (lastScene) {
        lastScene.leave();
      }

      if (outTransition === null) {
        outTransition = new FadeOutTransition(Color.black);
      }
      if (inTransition === null) {
        inTransition = new FadeInTransition(Color.black);
      }

      this.sceneDirector.pushScene(newScene, popLastScene, true, outTransition, inTransition);
    }
  }

  resetToMainMenu(): void {
    const mainMenuScene = this.sceneForSceneID(SceneID.MAINMENU);
    if (mainMenuScene) {
      this.sceneDirector.replaceStackWithScene(mainMenuScene);
    }
    this.game = new Game(this.game.getWorldMap());
  }

  // Save/load stubs for Phase 4
  serialize(_saveName: string): void {
    console.log('Save functionality not yet implemented - Phase 4');
  }

  deserialize(_saveName: string): Game | null {
    console.log('Load functionality not yet implemented - Phase 4');
    return null;
  }
}
